#10: Moved and updated TLS introduction. (#11)
exasol/tutorials: tls/doc/tls_introduction.md

# An Introduction to TLS

Before we dive into the Exasol-specifics of TLS connections in the [next article](tls_with_exasol.md), let’s first discuss some TLS basics. We’ve created this introduction for people who want to get an idea about how TLS works. I am using metaphors, and I will simplify a lot of the details to make the article more accessible for everyone. If you’re already a cryptography expert, you can skip directly to ["TLS With Exasol"](tls_with_exasol.md).

## What is TLS?

Transport Layer Security — or TLS — is the successor to the Secure Socket Layer (SSL). While being a [separate standard](https://tools.ietf.org/html/rfc5246), people have become so used to saying 'SSL' that they often use it as a synonym. For the sake of precision, we are going to stick with TLS in this article.

Simply put, TLS is meant to protect network communication against eavesdropping and offers a mechanism to verify that the other end of the connection is the one you intended to reach.

> Since people often have a wrong perception, I will stress here that a TLS connection does not protect against malicious intent of your communication partner.

Reviewer: ```suggestion
Since people often have an incorrect interpretation, I will stress here that a TLS connection does not protect against malicious intent of your communication partner.
```

The session layer is responsible for managing the connection state between communication partners. TLS has a session concept — so a part of TLS lives in this layer. The presentation layer is a translator between what the application expects and what the lower layers provide. In the case of TLS, this means encryption and decryption of application data.

The table above shows an example of an HTTPS connection. While you can see that the association between protocols and most layers is clear, TLS affects parts of both the session and presentation layer.

**TLS Session Layer (Layer 5) responsibilities:**

- Managing the TLS handshake process
- Establishing and maintaining secure sessions
- Negotiating cipher suites and protocol versions
- Handling session resumption and renegotiation
- Managing connection state between communication partners

**TLS Presentation Layer (Layer 6) responsibilities:**

- Encrypting outgoing application data
- Decrypting incoming encrypted data
- Data compression and decompression (when enabled)
- Converting between different data formats as needed

> This layered approach means that when Alice sends an HTTPS request to Bob's server, the application data first gets encrypted by TLS's presentation layer functions. While the session layer functions ensure the secure connection is properly maintained throughout the communication.

Reviewer: Not sure how to re-word to maintain content, but `While the session layer functions ensure the secure connection is properly maintained throughout the communication.` isn't a complete sentence currently.

"Supporting this, is the session layer which ensures the secure connection is maintained throughout the communication"

If Bob wants to digitally sign data, he first calculates a tamper-proof representation of that data using a cryptographic hash function. That proof is calculated from the contents of the data payload. The chosen algorithm does not allow reversing the process with any realistic effort. That means while Alice will be able to calculate representation from the payload, Eve can't create any meaningful data out of the number. This irreversible property is why cryptographic hash functions are also called 'one-way hash functions' — you can easily go from input to output, but not the other way around. Bob then uses his private key to encrypt the one-way hash and puts it alongside the payload data.

For practical reasons, the representation is, of course, small compared to the original data. It's communication overhead after all.

Alice, after receiving the payload and signature, can now calculate the hash on her own based on the payload data. She then uses Bob’s public key to decrypt the hash in the signature. If the hashes Alice and Bob calculated match, the payload data is accurate and consistent.

> Note that signatures only work towards the information security goal of integrity. Without any further measures they don’t do anything for confidentiality of the payload data.

Reviewer: ```suggestion
Note that signatures only work towards the information security goal of integrity. Without any further measures, they don’t do anything for confidentiality of the payload data.
```

### Certificates and Certification Agencies

Public key cryptography uses a chain of certificates to solve the problem of establishing the authenticity of communication partners. Certificates contain the public key and additional information that allows users to verify the key’s authenticity. They can be chained together so that the integrity of one certificate can be established using another one. At some point, though, there is no way around having a personal copy of a certificate you trust.

Alice uses a web browser to surf to Bob’s web page that is served via TLS. When she installed her computer, a lot of certificates were automatically installed too. Alice trusts the maker of that installation program to provide authentic certificates. These certificates were issued by a number of well-known organizations called Certification Agencies (CAs).

> Bob’s server presents its certificate to Alice’s client, and that client first checks whether it knows the certificate or not. The client realizes that it does not. Next, the client checks Bob’s server certificate for chaining information. If available, the client follows the chain down to a certificate that is installed on Alice’s machine. Luckily, Bob’s certificate is signed by ACME CA, a certification agency for which a certificate is pre-installed on the client machine. After it used, the key provided in the ACME CA certificate to check the signature, the client application tells Alice that the certificate for Bob’s server is valid. And, that the hostname of Bob’s web page and Bob’s certificate match. Alice can now securely surf Bob’s page.

Reviewer: Something in this clause at the start is off:
`After it used, the key provided in the ACME CA certificate to check the signature, ...`

There is a verb missing 'is' before used, but the clauses seem a bit out of order too. It seems fines  later on in the sentence.

It is perfectly possible to connect to a website that has a valid TLS certificate — and still get scammed on that site. In fact, that is pretty much the standard case for scams nowadays.

At the beginning of the World Wide Web, a couple of certification agencies claimed to add extra security on top of secure communication by verifying the certificate holder's identity and whether they are trustworthy. This is — with all due respect — a load of horse manure. Issuing certificates for a long time was a very profitable business model. Companies providing certificates advertised them as a means of proving that a web page can be trusted — a claim without substance.

Think about it:

* How does the company providing hundreds of thousands, or even millions, of certificates plan to reliably establish the identity of a certificate holder?
* How does the CA check the services are legitimate?
* If the issue holder’s services are now legitimate, how can the CA tell this will still be the case in a week or month?
* What if the issue holder uses the certificate to host someone else’s malicious service?
* What if a legit service was hijacked by an attacker?

> What the CAs really verify is that someone paid for their certificates. Preferably, with a credit card. While that might help law enforcement agencies to find the person behind a scam, it does nothing to protect you when you establish a TLS connection. You might find it consoling that some responsible people are put behind bars, but your data and potentially money are already gone by that time.

Reviewer: or you could use `responsible parties` as that has less of a nice connotation & is used more often with criminal activities

A TLS server certificate tells you whether the hostname of the server matches the one in the certificate. Period.

That’s its purpose, that’s what it does well, and it is also the only thing it does.

If you understood that fact, you will also understand why certificates issued by a non-profit organization like “Let’s Encrypt” are exactly as good or bad as one issued by Verisign.

### Client Certificates

One more part of authentication that we haven’t touched yet is authentication in the opposite direction. The reason for this is that in almost all cases servers don’t use TLS to establish the client's authenticity but mechanisms on the application layer like a login and a session ID or tokens.

TLS supports client certificates, but they are seldom used. Mostly because of the [too-many-keys problem](#the-too-many-keys-problem) we discussed earlier. Services typically do not trust public CAs enough to accept client certificates issued by someone outside their own organization.

> That being said, there are cases where client certificates are indeed useful. Mainly in machine-to-machine communication. IT departments often use their own CAs to issue client certificates that they hand out to developers of applications that access company services via TLS. The application then has to present its client certificate during the handshake phase. Only if the service can successfully authenticate the client via its client certificate, will it accept the TLS connection.

Reviewer: consistency and second clause not a complete sentence
```suggestion
That being said, there are cases where client certificates are indeed useful — mainly in machine-to-machine communication. IT departments often use their own CAs to issue client certificates that they hand out to developers of applications that access company services via TLS. The application then has to present its client certificate during the handshake phase. Only if the service can successfully authenticate the client via its client certificate will it accept the TLS connection.
```

### Truststores and Keystores

The words "keystore" and "truststore" are often used as synonyms. This is not precise. Truststores are a specific subset of keystores containing someone else’s public keys or certificates. If you install root CA certificates on a machine, they go into a truststore.

> In the Java world the distinction is very strict. When the official documentation says "keystore," it really only refers to the place where you store your own keys.

Reviewer: ok to keep, i think but wanted to ping as only language-specific use case so far

This behavior is also fueled by the fact that a lot of internet forum entries suggest setting the corresponding configuration options as a quick fix without explaining the dangers. The users are happy because their TLS connection now works and often do not realize that it is not secure anymore.

Why isn’t it secure? If you don’t check the validity of the certificate presented by a service, you create the perfect opportunity for a man-in-the-middle attack. This means the attacker does not even have to go through the trouble of trying to get a fake root CA certificate installed on your machine.

## TLS Versions and Cipher Suites

IT security evolves. Both the defensive part and the attacker side. Some mechanisms that once were an industry standard are now considered unsecure.

How can this happen?

> Remember when we said that cryptography is based on the idea that reversing it is needs to be hard enough? We now have a lot more computing power in our hands, and new mathematical algorithms still are invented that reduce the effort of the calculations involved from unrealistic to manageable. Also, the designers of security mechanisms are human and as such sometimes plant fundamental flaws in a design that make an update unavoidable.

Reviewer: `Remember when we said that cryptography is based on the idea that reversing it is needs to be hard enough?`
at first glance this sentence was hard to parse, i think that maybe the "it" & assumed noun antecedent conflict with the behavior we're describing.

`Remember when we said that purpose of encrypted data (or other appropriate participle noun combo) was to make it hard to do` or something like that. Sorry, don't have a clear suggestion. Think folks can also figure out what's meant too, so we can leave as is.

There is a long list of evolutionary steps that TLS and its predecessor SSL underwent that brought new hash algorithms, plugged design holes, or introduced new features required for new communication channels.

The designers of TLS knew that existing algorithms could not be secure forever, so TLS allows using a multitude of cipher suites in different versions. Those suites take care of signing, encryption, authentication, and validating integrity.

Client and server negotiate which suite and version to use by comparing which ones both support and picking the best available option.

> The important part here is to follow the announcements in the IT security field and make sure that all deployments use and accept only TLS versions and ciphers that are still considered hard enough. The TLS configuration on a modern machine allows you to forbid the client and server to use deprecated versions.

Reviewer: ```suggestion
The important part here is to follow the announcements in the IT security field and make sure that all deployments use and accept only TLS versions and ciphers that are still considered computationally robust enough. The TLS configuration on a modern machine allows you to forbid the client and server to use deprecated versions.
```

August 2025 TLS 1.3 is the industry standard, and older versions (including TLS 1.2) should not be used anymore!

Please take a look at OWASPS's ["Transport LayerSecurity Cheat Sheet"](https://cheatsheetseries.owasp.org/cheatsheets/Transport_Layer_Security_Cheat_Sheet.html) for an example of an advisory on picking the right setup.

## Summary

In this article, we discussed what TLS is and what you use it for. We learned about the difference between symmetric and asymmetric encryption and why the latter is only used during the phase in which the connection gets established. We covered certificates, keystores, and truststores in the context of authentication.

> In the next part we’ll build on this information and see where — and how — [TLS comes into play when using Exasol’s analytics database](tls_with_exasol.md).

Reviewer: ```suggestion
In the next part, we’ll build on this information and see where — and how — [TLS comes into play when using Exasol’s analytics database](tls_with_exasol.md).
```
